File: android-sound-end/app/src/main/java/com/kezry/dacdx/soundend/JitterBuffer.java
package com.kezry.dacdx.soundend;

import java.util.TreeMap;

public final class JitterBuffer {
    private final TreeMap<Long, byte[]> packets = new TreeMap<Long, byte[]>();
    private Long nextSequence;

    public synchronized void push(long sequence, byte[] payload) {
        packets.put(sequence, payload);
        if (nextSequence == null) {
            nextSequence = sequence;
        }
        notifyAll();
    }

    public synchronized byte[] pop(long timeoutMs) throws InterruptedException {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (packets.isEmpty()) {
            long waitMs = deadline - System.currentTimeMillis();
            if (waitMs <= 0) return null;
            wait(waitMs);
        }
        if (nextSequence != null && packets.containsKey(nextSequence)) {
            byte[] exact = packets.remove(nextSequence);
            nextSequence++;
            return exact;
        }
        if (packets.size() > 8) {
            Long first = packets.firstKey();
            byte[] payload = packets.remove(first);
            nextSequence = first + 1;
            return payload;
        }
        return null;
    }
}

File: windows-sender/src/Dacdx.Protocol/JitterBuffer.cs
namespace Dacdx.Protocol;

public sealed class JitterBuffer<T>
{
    private readonly SortedDictionary<uint, T> _packets = new();
    private uint? _nextSequence;

    public int Count => _packets.Count;

    public void Push(uint sequence, T packet)
    {
        _packets[sequence] = packet;
        if (_nextSequence is null || sequence < _nextSequence.Value)
        {
            _nextSequence = sequence;
        }
    }

    public bool TryPop(out T? packet)
    {
        packet = default;
        if (_nextSequence is null)
        {
            return false;
        }

        if (_packets.Remove(_nextSequence.Value, out var exact))
        {
            packet = exact;
            _nextSequence++;
            return true;
        }

        if (_packets.Count > 8)
        {
            var first = _packets.First();
            _packets.Remove(first.Key);
            _nextSequence = first.Key + 1;
            packet = first.Value;
            return true;
        }

        return false;
    }
}
